runtime/src/util/set_multimap.rs

```diff
@@ -1,126 +1,162 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::borrow::Borrow;
+use std::marker::PhantomData;
 
 use cid::Cid;
 use fvm_ipld_blockstore::Blockstore;
-use fvm_ipld_hamt::Error;
-use fvm_shared::clock::ChainEpoch;
-use fvm_shared::deal::DealID;
-use fvm_shared::HAMT_BIT_WIDTH;
+
+use crate::{ActorError, Config, Map2, MapKey};
 
 use super::Set;
-use crate::{make_empty_map, make_map_with_root, parse_uint_key, u64_key, Map};
+
+pub struct SetMultimapConfig {
+    pub outer: Config,
+    pub inner: Config,
+}
 
 /// SetMultimap is a hamt with values that are also a hamt but are of the set variant.
 /// This allows hash sets to be indexable by an address.
-pub struct SetMultimap<'a, BS>(pub Map<'a, BS, Cid>);
-
-impl<'a, BS> SetMultimap<'a, BS>
+pub struct SetMultimap<BS, K, V>
 where
     BS: Blockstore,
+    K: MapKey,
+    V: MapKey,
+{
+    outer: Map2<BS, K, Cid>,
+    inner_config: Config,
+    value_type: PhantomData<V>,
+}
+
+impl<BS, K, V> SetMultimap<BS, K, V>
+where
+    BS: Blockstore + Clone,
+    K: MapKey,
+    V: MapKey,
 {
     /// Initializes a new empty SetMultimap.
-    pub fn new(bs: &'a BS) -> Self {
-        Self(make_empty_map(bs, HAMT_BIT_WIDTH))
+    pub fn empty(bs: BS, config: SetMultimapConfig, name: &'static str) -> Self {
+        Self {
+            outer: Map2::empty(bs, config.outer, name),
+            inner_config: config.inner,
+            value_type: Default::default(),
+        }
     }
 
     /// Initializes a SetMultimap from a root Cid.
-    pub fn from_root(bs: &'a BS, cid: &Cid) -> Result<Self, Error> {
-        Ok(Self(make_map_with_root(cid, bs)?))
+    pub fn load(
+        bs: BS,
+        root: &Cid,
+        config: SetMultimapConfig,
+        name: &'static str,
+    ) -> Result<Self, ActorError> {
+        Ok(Self {
+            outer: Map2::load(bs, root, config.outer, name)?,
+            inner_config: config.inner,
+            value_type: Default::default(),
+        })
     }
 
     /// Retrieve root from the SetMultimap.
     #[inline]
-    pub fn root(&mut self) -> Result<Cid, Error> {
-        self.0.flush()
+    pub fn flush(&mut self) -> Result<Cid, ActorError> {
+        self.outer.flush()
     }
 
     /// Puts the DealID in the hash set of the key.
-    pub fn put(&mut self, key: ChainEpoch, value: DealID) -> Result<(), Error> {
+    pub fn put(&mut self, key: &K, value: V) -> Result<(), ActorError> {
         // Get construct amt from retrieved cid or create new
```

Comment:
```suggestion
        // Get construct hamt from retrieved cid or create new
```

this is a hamt inner isn't it?